linkerd/app/core/src/errors.rs

```diff
@@ -217,31 +219,60 @@ impl<RspB: Default + hyper::body::HttpBody> respond::Respond<http::Response<RspB
                     return Ok(rsp);
                 }
 
-                let status = http_status(&*error);
-                debug!(%status, version = ?self.version, "Handling error with HTTP response");
-                Ok(http::Response::builder()
+                let mut rsp = http::Response::builder()
                     .version(self.version)
-                    .status(status)
                     .header(http::header::CONTENT_LENGTH, "0")
                     .body(ResponseBody::default())
-                    .expect("error response must be valid"))
+                    .expect("error response must be valid");
+                let status = set_http_status(&*error, rsp.headers_mut());
+                *rsp.status_mut() = status;
+                debug!(%status, version = ?self.version, "Handling error with HTTP response");
+                Ok(rsp)
             }
         }
     }
 }
 
-fn http_status(error: &(dyn std::error::Error + 'static)) -> StatusCode {
-    if let Some(HttpError { http, .. }) = error.downcast_ref::<HttpError>() {
+fn set_http_status(
+    error: &(dyn std::error::Error + 'static),
+    headers: &mut http::HeaderMap,
+) -> StatusCode {
+    if let Some(HttpError { http, message, .. }) = error.downcast_ref::<HttpError>() {
+        headers.insert(L5D_HTTP_ERROR_MESSAGE, HeaderValue::from_static(message));
         *http
     } else if error.is::<ResponseTimeout>() {
+        headers.insert(
+            L5D_HTTP_ERROR_MESSAGE,
+            HeaderValue::from_static("request timed out"),
+        );
         http::StatusCode::GATEWAY_TIMEOUT
-    } else if error.is::<FailFastError>() || error.is::<tower::timeout::error::Elapsed>() {
+    } else if let Some(e) = error.downcast_ref::<FailFastError>() {
+        headers.insert(
+            L5D_HTTP_ERROR_MESSAGE,
+            HeaderValue::from_str(&e.to_string()).unwrap_or_else(|error| {
+                warn!(%error, "Failed to encode fail-fast error message");
+                HeaderValue::from_static("service in fail-fast")
+            }),
+        );
+        http::StatusCode::SERVICE_UNAVAILABLE
+    } else if error.is::<tower::timeout::error::Elapsed>() {
+        headers.insert(
+            L5D_HTTP_ERROR_MESSAGE,
+            HeaderValue::from_static("proxy dispatch timed out"),
```

Comment:
hmm, is this actually correct? the `tower::timeout` layer is used to implement both the dispatch timeout (which bounds the length of time requests spend in the proxy) _and_ the connect timeout, which is a different error. i'm not sure if all `tower::timeout::error::Elapsed` we could see here will actually be from the dispatch timeout.

i noticed that the metrics code _also_ assumes all `Elapsed` errors are from the dispatch timeout: https://github.com/linkerd/linkerd2-proxy/blob/00c3394ab36eeff14192e6ae52c41238376daabd/linkerd/app/core/src/errors.rs#L365-L366 i'm not sure if this is correct either --- @olix0r, do you happen to know if it's possible for the errors layer to see `Elapsed` errors generated by connect timeouts? if it is, i wonder if we should do something like wrapping the `tower` error type in a new error type to indicate whether an `Elapsed` error was due to a dispatch or connection timeout...

if we see connect timeouts here, i think they should *probably* get 504 Gateway Timeout instead of 503 Service Unavailable?

Reply:
By changing the test code to use `to_tcp_connect`, and changing it's signature to take a `connect_tcp`, the proxy code and test code can now both use this method to pass in the connect service that the stack should build off of.

In the test code, this `connect_tcp` is the [connect_timeout](https://github.com/linkerd/linkerd2-proxy/blob/911a4f03a6f73b854dac1c1a78a8560928495e1e/linkerd/app/inbound/src/http/tests.rs#L317) function that ensures the test proxy hits a connect timeout before a service is returned.

Additionally, `to_tcp_connect` now wraps timeout errors with a new `ConnectTimeout` which we can check for in [set_http_status](https://github.com/linkerd/linkerd2-proxy/blob/b41e85adb15fab891ef32c0330e44aa61656ff33/linkerd/app/core/src/errors.rs#L259-L264) and return a 504 Gateway Timeout instead of a 503 Service Unavailable.